### PR TITLE
AddFetpProject

### PR DESCRIPTION
**Pull request by nick-beyondessential**
### Changes:
Migration to add Project and dashbaord groups for FEPT

### CODE REVIEW

File: packages/database/src/migrations/20210118040638-AddFetpProject-modifies-data.js [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+      insertObject(db, 'dashboardGroup', {
+        organisationLevel: 'Project',
+        userGroup,
+        organisationUnitCode: projectCode,
+        dashboardReports: projectDashboardReports,
+        name: projectName,
+        code: getProjectDashboardGroupCode(userGroup),
+        projectCodes: `{${projectCode}}`,
+      }),
+    ),
```
Comment (billli0) [inline packages/database/src/migrations/20210118040638-AddFetpProject-modifies-data.js:72]: Is this a special case to add multiple `userGroup` in `orgnisationLevel: Project`? Correct me if I'm wrong, I guess the `userGroup` here usually is the most public one?

File: packages/database/src/migrations/20210118040638-AddFetpProject-modifies-data.js [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+};
+
+const countryCodes = ['PG', 'SB'];
+const projectCode = 'fetp';
+const projectName = 'FETP';
+const projectDescription = 'Field Epidemiologist Training Program Graduate Information';
+
+const dashboardGroupName = 'FETP';
+const projectDashboardReports = '{project_details}';
+const userGroups = ['Public', 'FETP Graduates'];
```
Comment (billli0) [inline packages/database/src/migrations/20210118040638-AddFetpProject-modifies-data.js:26]: From the ticket's description, I think there is another `userGroup` missing here, which is `FETP Admin`?

File: packages/database/src/migrations/20210118040638-AddFetpProject-modifies-data.js [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+const getProjectDashboardGroupCode = userGroup => `FETP_Project_${userGroup.split(' ').join('_')}`;
+
+const getCountryDashboardGroupCode = (countryCode, userGroup) =>
+  `${countryCode}_FETP_Country_${userGroup.split(' ').join('_')}`;
+
+export const hierarchyNameToId = async (db, name) => {
```
Comment (billli0) [inline packages/database/src/migrations/20210118040638-AddFetpProject-modifies-data.js:33]:
```suggestion
const hierarchyNameToId = async (db, name) => {
```

File: packages/database/src/migrations/20210118040638-AddFetpProject-modifies-data.js [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+};
+
+exports.down = async function (db) {
+  const hierarchyId = await hierarchyNameToId(db, projectCode);
+
+  await db.runSql(`DELETE FROM "dashboardGroup" WHERE name = '${projectName}'`);
+  await db.runSql(`DELETE FROM project WHERE code = '${projectCode}'`);
+  await db.runSql(`DELETE FROM entity_relation WHERE entity_hierarchy_id = '${hierarchyId}'`);
+  await db.runSql(`DELETE FROM entity_hierarchy WHERE name = '${projectCode}'`);
+  await db.runSql(`DELETE FROM entity WHERE code = '${projectCode}'`);
```
Comment (billli0) [inline packages/database/src/migrations/20210118040638-AddFetpProject-modifies-data.js:112]: Errow free deleting order👏

File: packages/database/src/migrations/20210118040638-AddFetpProject-modifies-data.js [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+      insertObject(db, 'dashboardGroup', {
+        organisationLevel: 'Country',
+        userGroup,
+        organisationUnitCode: countryCode,
+        dashboardReports: '{project_details}',
+        name: dashboardGroupName,
+        code: getCountryDashboardGroupCode(countryCode, userGroup),
+        projectCodes: `{${projectCode}}`,
+      }),
+    ),
```
Comment (billli0) [inline packages/database/src/migrations/20210118040638-AddFetpProject-modifies-data.js:56]: Here we can control the userGroup in `Country` level 👍

**Review by billli0 — CHANGES_REQUESTED** [review]
Thanks Nick, I've been learning from your very clean code👍. Leave a few comments for minor details.

File: packages/database/src/migrations/20210118040638-AddFetpProject-modifies-data.js (hunk 1 above)
Comment (nick-beyondessential) [reply packages/database/src/migrations/20210118040638-AddFetpProject-modifies-data.js:72]: Thanks @billli0, originally I was thinking from comment ' Basic project information can be available to the Public', that Public on project would be description dashboard and regional under 'FETP Graduates', but having a look through the cards looks like all reports are at country level, so I think you are correct and I'll limit project dashboard to public.

File: packages/database/src/migrations/20210118040638-AddFetpProject-modifies-data.js (hunk 2 above)
Comment (nick-beyondessential) [reply packages/database/src/migrations/20210118040638-AddFetpProject-modifies-data.js:26]: Looks like only 'Public', 'FETP Graduates' userGroups are mentioned in current cards, so we can add a Admin group when required.

**Review by billli0 — APPROVED** [review]
Looks good to me👍

**Pull request merged**